### PR TITLE
Support resolving ${env:...} in sourceFileMap

### LINKED ISSUE #3504: Debug configuration could be made more dynamic [sourceFileMap]
Type: Debugger
**Input information below**

**Describe the bug**
- OS and Version: `Microsoft Windows [Version 10.0.18875.1000]`
- VS Code Version:
```
Version: 1.33.1 (user setup)
Commit: 51b0b28134d51361cf996d2f0a1c698247aeabd8
Date: 2019-04-11T08:27:14.102Z
Electron: 3.1.6
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Windows_NT x64 10.0.18875
```
- C/C++ Extension Version:
```
Name: C/C++
Id: ms-vscode.cpptools
Description: C/C++ IntelliSense, debugging, and code browsing.
Version: 0.22.2-insiders
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools
```
- Other extensions you installed (and if the issue persists after disabling them): N/A
- A clear and concise description of what the bug is:

`Debug configuration` could be configured to be slightly more dynamic:
```json
"sourceFileMap": {"/mnt/c": "${env:systemdrive}/"}
```

**To Reproduce**
1. Open `.vscode/launch.json`
1. While in `C++ Launch`/`(gdb) Bash on Windows Launch`/ configuration scroll down to `sourceFileMap`
1. replace existing `sourceFileMap` with improved more dynamic one:
	```json
	"sourceFileMap": {
		"/mnt/c": "${env:systemdrive}/"
	}
	```
1. Unfortunately the `/mnt/c` could not be optimized further because using `$(wslpath $SYSTEMDRIVE/)` does not seem to work

### PR DESCRIPTION
**Pull request by WardenGnaw**
This PR resolves ${env:...} variables in sourceFileMap keys and values.

+ refactor of envFile

```
            "sourceFileMap":{
                "${env:windir}": "${env:driverdata}"
            }
```

Output Window:
image

Resolves: https://github.com/microsoft/vscode-cpptools/issues/3504

### CODE REVIEW

**Comment by pieandcakes** [comment]
did you test with unknown value and mix in more than one map?

File: Extension/src/Debugger/configurationProvider.ts [hunk 1]
```diff
@@ -252,6 +232,64 @@ class CppConfigurationProvider implements vscode.DebugConfigurationProvider {
...
+        }
+    }
+
+    private resolveSourceFileMapVariables(config: vscode.DebugConfiguration): void {
+        let messages: string[] = [];
+        if (config.sourceFileMap) {
+            for (const sourceFileMapKey of Object.keys(config.sourceFileMap)) {
+                const sourceFileMapValue: string = config.sourceFileMap[sourceFileMapKey];
+
+                // TODO: pass config.environment as initial environment to resolveVariables when it is { key: value } instead of { "key": key, "value": value }
```
Comment (pieandcakes) [inline Extension/src/Debugger/configurationProvider.ts]: should this be in the method above? this is talking about environment but you are in sourcefilemap

File: Extension/src/Debugger/configurationProvider.ts [hunk 2]
```diff
@@ -252,6 +232,64 @@ class CppConfigurationProvider implements vscode.DebugConfigurationProvider {
         return config && config.type ? config : null;
     }
 
+    private convertEnvFileToEnvironmentMap(config: vscode.DebugConfiguration, folder: vscode.WorkspaceFolder): void {
```
Comment (pieandcakes) [inline Extension/src/Debugger/configurationProvider.ts]: Should this be called resolveEnvFile ? because you are resolving it to the environment object.

**Review by pieandcakes — APPROVED** [review]

File: Extension/src/Debugger/configurationProvider.ts (hunk 1 above)
Comment (WardenGnaw) [reply Extension/src/Debugger/configurationProvider.ts]: No. This is for using config.environment variables in sourceFileMap.

E.g.
environment: {
   "abc": "efg"
},
sourceFileMap: {
   ${env:abc}: ...
}

**Review by pieandcakes — APPROVED** [review]

**Pull request merged**